models/pull.go

```diff
@@ -1402,3 +1402,37 @@ func TestPullRequests() {
 func InitTestPullRequests() {
 	go TestPullRequests()
 }
+
+// CloseActivePullRequests closes the issue of all the active pull requests associated with a branch
+func CloseActivePullRequests(doer *User, repo *Repository, branchName string) {
+	baseprs, err := GetUnmergedPullRequestsByBaseInfo(repo.ID, branchName)
+	if err != nil {
+		log.Error(4, "Find pull requests [base_repo_id: %d, base_branch: %s]: %v", repo.ID, branchName, err)
+		return
+	}
+	headprs, err := GetUnmergedPullRequestsByHeadInfo(repo.ID, branchName)
+	if err != nil {
+		log.Error(4, "Find pull requests [head_repo_id: %d, head_branch: %s]: %v", repo.ID, branchName, err)
+		return
+	}
+	prs := append(baseprs, headprs...)
+
+	for _, pr := range prs {
+		err = pr.LoadIssue()
+		if err != nil {
+			log.Error(4, "LoadIssue: %v", err)
+			return
+		}
+		err = pr.Issue.ChangeStatus(doer, true)
+		if err != nil {
+			if IsErrDependenciesLeft(err) {
+				log.Warn("Issue status change blocked by existing dependencies: %v", err)
+			} else {
+				log.Error(4, "ChangeStatus: %v", err)
+			}
+		} else {
+			log.Trace("Issue [%d] status changed to closed", pr.IssueID)
+			go CreateOrUpdateIssueNotifications(pr.Issue, doer.ID)
```

Comment:
I don't think this should be needed

Reply:
So watchers shouldn't be notified when a PR gets closed?

Reply:
I can't directly call notifications method because it creates a cyclic dependency.

Reply:
@HarshitOnGitHub @lafriks is right. This line is unnecessary because https://github.com/go-gitea/gitea/blob/3a3c9f17edc47959fa0504e2596365bf9e958b80/models/issue_comment.go#L632 has done that. The code is confusing and so I have sent serval refactors about notifications but not reviewed and merged yet. See #5110 #5111 #5113.

Reply:
Oh! @lunny Thanks for the explanation! @lafriks I have updated the PR. :)